Don't 500 on missing pendingAuthz

cpu opened the pull request:
As described in #3201  a concurrent challenge POST would result in 500 errors if the pending authz row was deleted by a promotion to the authz table underneath another request. 

This PR adjusts the SA & RA so that if a pending authz is promoted to a final authz between updates a not found error will be returned instead of a server internal error.

Update: Removed mentions of the additional lockcol problem. The root cause of this is still TBD so we've split #3201 / this PR to only address the missing pending authz.

Resolves https://github.com/letsencrypt/boulder/issues/3201

jsha commented on ra/ra_test.go:
```diff
@@ -2247,6 +2247,42 @@ func TestFinalizeOrder(t *testing.T) {
 }
 
+// TestUpdateMissingAuthorization tests the race condition where a challenge is
+// updated to valid concurrent to another attempt to have the challenge updated.
```
*to valid concurrently with another ...

(just reads a little nicer)

jsha commented on sa/_db-next/migrations/20171120132231_DropChallengesLockCol.sql:
```diff
@@ -0,0 +1,11 @@
+-- +goose Up
+-- SQL in section 'Up' is executed when this migration is applied
+
+ALTER TABLE challenges DROP COLUMN LockCol;
```
I would leave out this migration. I don't think we are expecting to get the challenges table down to a size where we can migrate it any time soon. I think it's sufficient to leave the LockCol in place but unused.

cpu replied:
Ok. It's unpleasant leaving unused DB fields laying around but there also doesn't seem to be a chance we can land a migration so I think my hands are tied.

jsha commented on sa/sa.go:
```diff
@@ -139,10 +139,28 @@ func updateChallenges(authID string, challenges []core.Challenge, tx *gorp.Trans
+		results, err := tx.Exec(
+			`UPDATE challenges
+			SET status = ?,
+			error = ?,
```
Indent these lines to group them more clearly.

jsha commented on sa/sa.go:
```diff
@@ -139,10 +139,28 @@ func updateChallenges(authID string, challenges []core.Challenge, tx *gorp.Trans
-		_, err = tx.Update(chall)
+
+		// We use an explicit UPDATE here with a set of fields we care about instead
+		// of using gorp's `tx.Update()` to avoid the LockCol and concurrent update
```
I think in the future, when LockCol doesn't exist, this sentence won't make as much sense. Instead it would be something like "instead of using `tx.Update()` because we are only modifying one column."

Reading through this again, I realize there's another reason we don't need a LockCol on challenges: We only modify the challenges within a transaction. In other words, we were using opportunistic locking at the same time as regular locking! This means that the existing `tx.Update(chall)` is safe enough even without the column-specific UPDATE here. I think I would opt to keep it the same, though I'm open to thoughts otherwise!

Sorry I gave you bad advice earlier about the column-specific update; I hadn't checked this code, and had forgotten that we hold a transaction while we do it.

cpu replied:
> Reading through this again, I realize there's another reason we don't need a LockCol on challenges: We only modify the challenges within a transaction.

Hmm. That's an interesting point but leaves me a little bit confused. How would the opportunistic lock col error be happening if the concurrent UPDATES are inside of a transaction? Wouldn't the isolation properties result in each incrementing the LockCol appropriately after waiting for the transaction in-flight to complete? 

> I think I would opt to keep it the same, though I'm open to thoughts otherwise!

"the same" -> As an explicit `UPDATE`? Or "the same" -> the same as it is in master, a `tx.Update()`?

> Sorry I gave you bad advice earlier about the column-specific update; I hadn't checked this code, and had forgotten that we hold a transaction while we do it.

No worries. I should have thought about it a little bit more deeply when I was looking at the transactions.

cpu commented:
I've rebased this PR & edited its description to match the update of #3201 to cover only the missing pending authorization case. The gorp lock error is now split out into a separate issue: #3249

@jsha @rolandshoemaker This is ready for review again. Thanks!
